jq255.md

```diff
@@ -0,0 +1,1412 @@
+
+  - `is_square(x)` returns `True` if `x` is a square in the finite field,
+    or `False` otherwise.
+
+  - `scalar_to_int(s)` and `int_to_scalar(j)` compute the same
+    operations as `int()` and `GFq()`, respectively, but for scalars,
+    which are integers modulo the prime `r`.
+
+  - `secure_random(m)`, for an integer `m >= 1`, returns a sequence of
+    `m` random bytes obtained from a cryptographically secure source,
+    with a uniform probability distribution (e.g. `/dev/urandom` or the
+    `getrandom()` system call on many Unix-like operating syustems).
+    This is used only for private key generation in the specification of
+    high-level protocols.
+
+If an addition, subtraction or multiplication uses as operands a field
+element and an integer (usually small), then the integer is first
+implicitly converted to a field element as if by using `GFq()`. In
+practice, the implementation may provide a dedicated
+multiply-by-small-integer function; on usual software platforms, such an
+operation is about as fast as an addition in the field.
+
+The implementation of `sqrt(x)` depends on the used field. For jq255s,
+the field is such that `q = 3 mod 4`, which allows a simple square root
+implementation by computing the square root candidate `z = x^((q+1)/4)`.
+For jq255e, the field is such that `q = 5 mod 8`, which can be handled
+with Atkin's algorithm:
+
+ 1. Compute: `c = (2*x)^((q-5)/8)`
+ 2. Compute: `d = 2*x*c^2`
+ 3. Compute the square root candidate: `z = x*c*(d - 1)`
+
+The obtained candidate MUST be checked (by comparing `z^2` with `x`).
+Moreover, if `z` is indeed a square root of `x`, then it MUST be
+normalized to a non-negative value (i.e. if `sgn(z) = 1`, then `z` must
+be replaced with `-z`). The use of this fixed convention for square
+roots is needed for proper interoperability of the map-to-group and
+hash-to-group operations.
+
+`is_square(x)` can be implemented as a simple wrapper around `sqrt(x)`,
+though faster implementations are possible with the [Legendre/Jacobi
+symbol][]. A fast, constant-time implementation of that operation can be
+obtained as a straightforward derivative of the [optimized binary
+GCD][]; on some small microcontrollers, this `is_square()`
+implementation can be about six times faster than the full square root
+computation.
+
+### Affine Formulas
+
+We recall here the affine formulas applicable to points on double-odd
+Jacobi quartic curves. Implementations are expected to use the extended
+representation described later on in this document; the affine formulas
+are potentially useful if some alternate point representations are
+explored.
+
+For any point `P = (e, u)` on the curve, the point `P+N` (i.e. the other
+representant of the same group element) is `P+N = (-e, -u)`. The
+*opposite* of the point `P` is `-P = (e, -u)`. The neutral element of
+the group is `{ I, N }` (i.e. `I = (1, 0)` and `N = (-1, 0)` are the two
+representants of the group neutral).
```

Comment:
Is the *opposite* the inverse?

Reply:
If `N = (-1, 0)` and `P=(e, u)`, how does `P+N = (-e, -u)`?

Reply:
I think "opposite" is the correct term if we talk about "adding" points; "inverse" relates to multiplication, but here I am denoting the group law additively.

Reply:
If you start with `P = (e,u)`, you have `-P = (e, -u)`, `P+N = (-e, -u)` and `-P+N = (-e, u)`. The point `N` itself is `(-1, 0)`, hence `N+N = I = (1, 0)` (since `N` has order 2, adding it to itself yields the point-that-was-at-infinity).

In the Weierstraß curve, `u` is the inverse of the slope of the line that goes from `N` to `P`, and that line (by construction) also intersects the curve at point `-(P+N) = -P+N`, so it is expected that `P` and `-P+N` share the same `u`, while point negation is symmetry relatively to the horizontal, hence `-P` must have the opposite slope (hence `-u`). Geometric interpretation of `e` is a bit harder in the Weierstraß domain, because `e` is "unscaled" (curve isomorphisms are obtained by multiplying `x` by `d^2`, `y` by `d^3`, `a` by `d^2` and `b` by `d^4` for any non-zero `d`; but `e` is unchanged through such isomorphisms).